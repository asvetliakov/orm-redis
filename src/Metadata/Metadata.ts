/**
 * Metadata key to mark class to be stored as redis hash
 */
export const REDIS_ENTITY = Symbol("Redis Hash");

/**
 * Metadata key to store redis properties metadatas
 */
export const REDIS_PROPERTIES = Symbol("Redis properties");

/**
 * Metadata key to keep serialized initial value of property from redis. For collections this will be link to the corresponding hashmap/set
 */
export const REDIS_VALUE = Symbol("Redis initial value");

/**
 * Metadata key to keep the initial state of redis collections (map/set)
 */
export const REDIS_COLLECTION_VALUE = Symbol("Redis initial collection value");


export type PropertyMetadata = SimplePropertyMetadata | RelationPropertyMetadata;

/**
 * Hash simple property metadata
 * 
 * @export
 * @interface SimplePropertyMetadata
 */
export interface SimplePropertyMetadata {
    /**
     * Property name in class
     */
    propertyName: string;
    /**
     * Property name stored in redis
     */
    propertyRedisName: string;
    /**
     * Property type
     */
    propertyType: any;
    /**
     * True if property must be identify column
     */
    isIdentifyColumn: boolean;
    /**
     * Relation flag
     */
    isRelation: false;
}

/**
 * Relation options
 * 
 * @export
 * @interface RelationOptions
 */
export interface RelationOptions {
    /**
     * Property name in redis
     */
    propertyName?: string;
    /**
     * True to automatically create new hash for relation
     */
    cascadeInsert?: boolean;
    /**
     * True to automatically update corresponding relation hash
     */
    cascadeUpdate?: boolean;
}

/**
 * Hash relation property metadata
 * 
 * @export
 * @interface RelationPropertyMetadata
 */
export interface RelationPropertyMetadata {
    /**
     * Property name in class
     */
    propertyName: string;
    /**
     * Property name stored in redis
     */
    propertyRedisName: string;
    /**
     * Property type.
     * For set property relation it will be Set itself
     * For single property relation it will be equal to relationType
     */
    propertyType: any;
    /**
     * Relation flag
     */
    isRelation: true;
    /**
     * Relation type. Must be type of related entity
     */
    relationType: any;
    /**
     * Relation options
     */
    relationOptions: RelationOptions;
}

/**
 * Return true if given object is redis entity like
 * 
 * @export
 * @param entity 
 * @returns 
 */
export function isRedisEntity(entity: object): boolean {
    if (typeof entity !== "object" || entity === null) {
        return false;
    }
    const metadata = Reflect.getMetadata(REDIS_ENTITY, entity.constructor);
    return !!metadata;
}

/**
 * Return entity name
 * 
 * @export
 * @param entityOrEntityClass 
 * @returns 
 */
export function getEntityName(entityOrEntityClass: object | Function): string | undefined {
    const metadata = Reflect.getMetadata(REDIS_ENTITY, typeof entityOrEntityClass === "function" ? entityOrEntityClass : entityOrEntityClass.constructor);
    return metadata;
}

/**
 * Return entity id
 * 
 * @export
 * @param entity 
 * @returns 
 */
export function getEntityId(entity: { [key: string]: any }): string | number | undefined {
    if (!isRedisEntity(entity)) {
        return undefined;
    }
    const metadatas: PropertyMetadata[]  = Reflect.getMetadata(REDIS_PROPERTIES, entity.constructor);
    const idMetadata = metadatas && metadatas.find(val => !val.isRelation && val.isIdentifyColumn);
    if (!idMetadata) {
        return undefined;
    }
    if (typeof entity[idMetadata.propertyName] === "number" || typeof entity[idMetadata.propertyName] === "string") {
        return entity[idMetadata.propertyName];
    }
    return undefined;
}

/**
 * Returns redis properties metadata
 * 
 * @export
 * @param entityOrEntityClass 
 * @returns 
 */
export function getEntityProperties(entityOrEntityClass: object | Function): PropertyMetadata[] | undefined {
    return Reflect.getMetadata(REDIS_PROPERTIES, typeof entityOrEntityClass === "function" ? entityOrEntityClass : entityOrEntityClass.constructor);
}

/**
 * Return full hash id
 * 
 * @export
 * @param entityOrEntityClass 
 * @param [hashId]
 * @returns 
 */
export function getEntityFullId(entityOrEntityClass: object | Function, hashId?: string | number): string | undefined {
    const name = getEntityName(entityOrEntityClass);
    const id = typeof hashId !== "undefined"
        ? hashId
        : typeof entityOrEntityClass === "object"
            ? getEntityId(entityOrEntityClass)
            : undefined;        
    if (name && typeof id !== "undefined") {
        return `e:${name}:${id}`;
    }
}