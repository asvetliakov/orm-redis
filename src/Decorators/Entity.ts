import { REDIS_ENTITY } from "../Metadata/Metadata";

/**
 * Defines redis entity class
 * 
 * @export
 * @param {string} [name] 
 * @returns void
 */
export function Entity(name?: string): ClassDecorator {
    return function (constructor: Function) {
        Reflect.defineMetadata(REDIS_ENTITY, name ? name : constructor.name, constructor);
    };
}