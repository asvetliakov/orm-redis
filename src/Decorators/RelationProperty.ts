import { MetadataError } from "../Errors/Errors";
import { PropertyMetadata, REDIS_PROPERTIES, RelationOptions, RelationTypeFunc } from "../Metadata/Metadata";

/** 
 * default options
 */
const defaultOptions: RelationOptions = {
    cascadeInsert: false,
    cascadeUpdate: false
};

/**
 * Defines single or multiple relation property
 * 
 * @export
 * @param type Type function which must return type of relation / property type
 * @param propertyType Property type. For single relation must be 
 */
export function RelationProperty(type: RelationTypeFunc): PropertyDecorator;
/**
 * Defines single or multiple relation property
 * 
 * @export
 * @param type Type function which must return type of relation / property type
 * @param [options] Options
 */
export function RelationProperty(type: RelationTypeFunc, options?: RelationOptions): PropertyDecorator;

/**
 * Defines single or multiple relation property
 * 
 * @export
 * @param type 
 * @param [options] 
 * @returns 
 */
export function RelationProperty(type: RelationTypeFunc, options?: RelationOptions): PropertyDecorator {
    return function (target: Object, propertyKey: string): void {
        const designType = Reflect.getMetadata("design:type", target, propertyKey);
        
        const redisPropertyName = options && options.propertyName ? options.propertyName : propertyKey;
        
        const relationTypes = type();

        let propertyType: Function;
        if (Array.isArray(relationTypes)) {
            [, propertyType] = relationTypes;
        } else {
            if (designType === Object) {
                throw new MetadataError(target.constructor, `Relation's ${propertyKey} property type is detected as simple Object. This is error. Specify property type explicitly`);
            }
            propertyType = designType;
        }

        const finalOptions = {
            ...defaultOptions,
            ...options
        };

        const properties: PropertyMetadata[] = Reflect.getMetadata(REDIS_PROPERTIES, target.constructor) || [];
        properties.push({
            isRelation: true,
            propertyName: propertyKey,
            propertyRedisName: redisPropertyName,
            propertyType: propertyType,
            relationTypeFunc: type,
            relationOptions: finalOptions
        });
        Reflect.defineMetadata(REDIS_PROPERTIES, properties, target.constructor);
    };
}