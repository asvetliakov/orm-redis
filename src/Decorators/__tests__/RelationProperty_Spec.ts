import { PropertyMetadata, REDIS_PROPERTIES, RelationPropertyMetadata } from "../../Metadata/Metadata";
import { Entity } from "../Entity";
import { RelationProperty } from "../RelationProperty";
@Entity()
class Rel { }

it("With default values", () => {
    @Entity()
    class C {
        @RelationProperty(() => Rel)
        public test: Rel;
    }
    const metadata: PropertyMetadata[] = Reflect.getMetadata(REDIS_PROPERTIES, C);
    expect(metadata).toMatchSnapshot();
    const relMetadata: RelationPropertyMetadata = metadata[0] as RelationPropertyMetadata;
    expect(relMetadata.relationTypeFunc()).toBe(Rel);
    expect(relMetadata.propertyType).toBe(Rel);
});

it("With relation options", () => {
    @Entity()
    class C {
        @RelationProperty(type => Rel, { cascadeUpdate: true, propertyName: "testName" })
        public test: Rel;
    }
    const metadata: PropertyMetadata[] = Reflect.getMetadata(REDIS_PROPERTIES, C);
    expect(metadata).toMatchSnapshot();
});

it("With explicitly specified property type", () => {
    @Entity()
    class C {
        @RelationProperty(() => [Rel, Rel])
        public test: Rel;
    }
    const metadata: PropertyMetadata[] = Reflect.getMetadata(REDIS_PROPERTIES, C);
    const relMetadata: RelationPropertyMetadata = metadata[0] as RelationPropertyMetadata;
    expect(relMetadata.relationTypeFunc()).toEqual([Rel, Rel]);
    expect(relMetadata.propertyType).toBe(Rel);
});

it("For set or map of relations", () => {
    @Entity()
    class C {
        @RelationProperty(() => [Rel, Set])
        public test: Set<Rel>;

        @RelationProperty(() => [Rel, Map])
        public test2: Map<string, Rel>;
    }
    const metadata: PropertyMetadata[] = Reflect.getMetadata(REDIS_PROPERTIES, C);
    expect(metadata).toMatchSnapshot();
    let relMetadata: RelationPropertyMetadata = metadata[0] as RelationPropertyMetadata;
    expect(relMetadata.relationTypeFunc()).toEqual([Rel, Set]);
    expect(relMetadata.propertyType).toBe(Set);
    relMetadata = metadata[1] as RelationPropertyMetadata;
    expect(relMetadata.relationTypeFunc(0)).toEqual([Rel, Map]);
    expect(relMetadata.propertyType).toBe(Map);
});