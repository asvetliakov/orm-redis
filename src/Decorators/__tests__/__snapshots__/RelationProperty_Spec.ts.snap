// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`For set or map of relations 1`] = `
Array [
  Object {
    "isRelation": true,
    "propertyName": "test",
    "propertyRedisName": "test",
    "propertyType": [Function],
    "relationOptions": Object {
      "cascadeInsert": false,
      "cascadeUpdate": false,
    },
    "relationTypeFunc": [Function],
  },
  Object {
    "isRelation": true,
    "propertyName": "test2",
    "propertyRedisName": "test2",
    "propertyType": [Function],
    "relationOptions": Object {
      "cascadeInsert": false,
      "cascadeUpdate": false,
    },
    "relationTypeFunc": [Function],
  },
]
`;

exports[`With default values 1`] = `
Array [
  Object {
    "isRelation": true,
    "propertyName": "test",
    "propertyRedisName": "test",
    "propertyType": [Function],
    "relationOptions": Object {
      "cascadeInsert": false,
      "cascadeUpdate": false,
    },
    "relationTypeFunc": [Function],
  },
]
`;

exports[`With relation options 1`] = `
Array [
  Object {
    "isRelation": true,
    "propertyName": "test",
    "propertyRedisName": "testName",
    "propertyType": [Function],
    "relationOptions": Object {
      "cascadeInsert": false,
      "cascadeUpdate": true,
      "propertyName": "testName",
    },
    "relationTypeFunc": [Function],
  },
]
`;
